WIP: OpenEMR minimum web technologies requirements

MatthewVita opened the pull request.

MatthewVita commented on setup.php:
```diff
@@ -143,13 +143,24 @@ function cloneClicked() {
 <span class="text">
 
 <?php
+// Checks if the server's PHP version and the client's browser are compatible with OpenEMR:
```
@bradymiller I only am using this in the one file for now as an example. I'm interested in your feedback before copy/pasting it elsewhere :)

MatthewVita commented on common/compatibility/checker.php:
```diff
@@ -0,0 +1,103 @@
+    /**
+     * This is needed per the modernization project.
+     */
+    private $minimumPhpVersion = "5.5.9";
```
I'm pretty we have to hard code this because globals won't work out of the OpenEMR context, right?

bradymiller replied:
yep, which is fine

bradymiller replied:
btw, the current minimum support php version is now 5.4.0

MatthewVita commented on common/compatibility/checker.php:
```diff
@@ -0,0 +1,103 @@
+     * aren't specified because these browsers tend to self
+     * update/encourage end users to stay current.
+     */
+    private $supportedBrowsers = array("Firefox",
```
I'm pretty we have to hard code this because globals won't work out of the OpenEMR context, right?

bradymiller replied:
yep, which is fine

bradymiller replied:
as a strategy, will be good to only send a warning here for browsers(on just the login screen; maybe just a warning at the bottom of the screen) rather than die since many browsers will likely work and/or folks may want to at least test it out on other browsers.

bradymiller replied:
disregard above comment, I like what you do with the alert popup mechanism for both issues

MatthewVita replied:
Okay, so the alert is only happening on:
- `setup.php`
- `sql_upgrade.php`
- `sql_patch.php`
- `acl_upgrade.php`
- `admin.php`
- `interface/login/login.php`

Note that login is included here. Figured that it would be good to show a message on the login page for end users before they start using OpenEMR.

Thoughts?

Also, I didn't include globals.php... don't know if it needs to go in there.

bradymiller replied:
Agree with login.php.
Agree with not doing globals.php since there are cases where the user will wish to press ahead despite the warning, and I could not imagine the frustration of getting that alert on every script :)
(plus globals.php does not always have access to the gui, for example in ajax scripts)

MatthewVita commented on common/compatibility/checker.php:
```diff
@@ -0,0 +1,103 @@
+                $response .= "\\n\\n";
+            }
+
+            $response .= "Your browser needs to be one of the following: ";
```
Maybe I'm just being lazy, but it seems like the setup and upgrade files where this code will be used doesn't make use of translations so...

bradymiller replied:
In this case, check for the existence of the xlt() function. If it doesn't exist, then can create a dummy xlt() function on the fly that does nothing and  then run the message through the xlt() function.

bradymiller replied:
Actually, don't use xlt(), use the xl() function instead.

bradymiller replied:
On more thought, rec not creating an xl() on the fly since it may be run prior to the official xl function in the login screen. So, if it exists, then would use the function, and if it doesn't exist, then don't use the function.(this may mean making duplicative lines of code here, but minor issue)

MatthewVita replied:
Check out the latest commit amend to this PR. I have a delegate function to use `xl` if it exists. Looking forward to your feedback.

bradymiller commented on common/compatibility/checker.php:
```diff
@@ -0,0 +1,103 @@
+ * @author  Matthew Vita <matthewvita48@gmail.com>
+ * @link    http://www.open-emr.org
+ */
+
```
Won't this need a namespace?

MatthewVita replied:
We don't need a namespace here because this class is mostly required outside of the OpenEMR context. I added a comment about this in the class.

MatthewVita replied:
TL;DR: won't be using autoloading to bring in the class.

bradymiller commented on setup.php:
```diff
@@ -143,13 +143,24 @@ function cloneClicked() {
+if (!empty($potentialWarnings)) {
+  echo "<script>alert('" . $potentialWarnings . "');</script>";
+}
+?>
```
I think this is good, since it allows the user to then go at their own demise, if they wish :)

bradymiller commented on setup.php:
```diff
@@ -143,13 +143,24 @@ function cloneClicked() {
+$compatCheck = new Checker();
+$potentialWarnings = $compatCheck->getCompatibility();
+if (!empty($potentialWarnings)) {
+  echo "<script>alert('" . $potentialWarnings . "');</script>";
```
run this through addslashes to escape:
``````
cho "<script>alert('" . addslashes($potentialWarnings) . "');</script>";

MatthewVita replied:
This messes up the \\n\\n line break. Is this really needed since we control the string that will be shown in the alert? In other words, it wouldn't be likely that naughty characters would get into this alert.

bradymiller commented:
Review complete :)

bradymiller commented on setup.php:
```diff
@@ -143,13 +143,24 @@ function cloneClicked() {
 
 <?php
+// Checks if the server's PHP version and the client's browser are compatible with OpenEMR:
+require_once(dirname(__FILE__) . "/common/compatibility/checker.php");
```
As an aside, can we just include the composer autoloader script on these scripts(setup, upgrades, and admin) without context?

bradymiller replied:
skip the above idea also :)
Would likely just cause pain (for example, it would define a broken xl() function)

MatthewVita replied:
Ok.

mdsupport commented:
Does adding a try/catch block for entire globals code work?  Catch block should return javascript alert and die.

MatthewVita commented:
@mdsupport This would most like work, I just don't know if we this shown on every page. I asked Brady about this above.

bradymiller commented on interface/login/login.php:
```diff
@@ -108,6 +108,17 @@ function imsubmitted() {
+$compatCheck = new Checker();
+$potentialWarnings = $compatCheck->getCompatibility();
+if (!empty($potentialWarnings)) {
+  echo "<script>alert('" . $potentialWarnings   . "');</script>";
```
agree on not using the addslashes (didn't realize it would break \n)

bradymiller commented:
@mdsupport ,
regarding placing this check and alert in globals.php. check out my comment I just placed regarding this in the review.

bradymiller commented:
@mdsupport, Quick question regarding the try/catch block in globals, which seems like it would be separate from this checking class, but is intriguing, and could be another layer when there is a real error.
@MatthewVita , what are your thoughts on this as a separate addition?(would be another PR)

bradymiller commented on acl_upgrade.php:
```diff
@@ -113,6 +113,15 @@
+$potentialWarnings = $compatCheck->getCompatibility();
+if (!empty($potentialWarnings)) {
+  echo "<script>alert('" . $potentialWarnings   . "');</script>";
+}
```
Shouldn't we always place this check prior to the globals.php include?

bradymiller commented on common/compatibility/checker.php:
```diff
@@ -0,0 +1,117 @@
+
+            $response .= xlDelegate("Your browser needs to be one of the following") . ": ";
+            $response .= "[" . implode(", ", $this->supportedBrowsers) . "] ";
+            $response .= xlDelegate("The OpenEMR development team exclusively tests " .
```
Shouldn't break lines. Each line goes as a separate constant for the translators, so best to keep them as one line (even if it is a paragraph of text).
``````
xlDelegate("The OpenEMR development team exclusively tests with modern versions of these supported browsers")...

bradymiller commented on common/compatibility/checker.php:
```diff
@@ -0,0 +1,117 @@
+     * @return empty string or string with compatability warnings.
+     */
+    public function getCompatibility() {
+        function xlDelegate($value) {
```
Will need to also make a note(probably best to make an Issue for it in github) to capture these strings in the constant collector perl script:
https://github.com/openemr/openemr/blob/master/contrib/util/language_translations/collectConstants.pl#L153-L165

MatthewVita commented:
@bradymiller I didn't forget about this, btw. Will return to it this week sometime.

MatthewVita commented:
@bradymiller I'm setting up an older version of Ubuntu with a PHP version less than 5.4.0... going to test with @mdsupport's try/catch strategy. If it works, I will rework this PR entirely :)

MatthewVita commented:
...still trying to get an environment with PHP 5.3 working. Using this `docker-compose.yml`, but I'm not quite there yet:

```
version: '2'

services:
   db:
     image: mysql:5.7
     volumes:
       - db_data:/var/lib/mysql
     restart: always
     environment:
       MYSQL_ROOT_PASSWORD: root
       MYSQL_DATABASE: openemr
       MYSQL_USER: root
       MYSQL_PASSWORD: pass

   apacheandphp:
     depends_on:
       - db
     image: bylexus/apache-php53
     ports:
       - "8000:80"
     restart: always
     volumes:
        - /var/www/openemr:/var/www/openemr
volumes:
    db_data:
```

MatthewVita commented:
Okay, @mdsupport and @bradymiller I finally got back to this PR. Here is my report.

I was able to finally get an old version of PHP 5.3 installed via Docker:
```
version: '2'

services:
   db:
     image: mysql:5.7
     volumes:
       - db_data:/var/lib/mysql
     restart: always
     ports:
       - "3306:3306"
     environment:
       MYSQL_ROOT_PASSWORD: 'pass'
       MYSQL_DATABASE: 'openemr'
       MYSQL_USER: 'root'
       MYSQL_PASSWORD: 'pass'

   apacheandphp:
     depends_on:
       - db
     image: bylexus/apache-php53
     ports:
       - "8000:80"
     restart: always
     volumes:
        - /var/www/openemr:/var/www/openemr
volumes:
    db_data:
```

I've simplified this PR to just check for PHP version... this is the message one receives on setup.php, sql_upgrade.php, sql_patch.php, acl_upgrade.php, admin.php, and globals.php:

![image](https://cloud.githubusercontent.com/assets/1715206/23103338/11af5f5a-f688-11e6-83d4-a1576d5a902d.png)

The browser check has been removed. I just updated the note about this in the setup.php:

```
<li>To ensure a consistent look and feel throughout the application,
     <a href='http://www.mozilla.org/products/firefox/'>Firefox</a> and <a href="https://www.google.com/chrome/browser/desktop/index.html">Chrome</a> are recommended. The OpenEMR development team exclusively tests with modern versions of these browsers.</li>
```

Probably good enough.

@bradymiller can you test these changes out?

Thanks,
Matthew

bradymiller commented:
@MatthewVita ,
Code looks really good. Will bring this in. A nice test of this will be in the error logs of the up for grabs demos, which I will check tomorrow. Thanks for the contribution!
Will be good to document somewhere on the wiki (for developers, especially) this minimum php version and where we set it in the codebase in common/compatibility/checker.php.
-brady

MatthewVita commented:
Good call Brady. I added a note here: http://www.open-emr.org/wiki/index.php/Development_Policies#PHP

Thanks!

bradymiller commented:
thanks for doing that. Also set it here, so we think about it every release:
http://www.open-emr.org/wiki/index.php/Steps_for_an_official_release#3._Publish_installation.2C_upgrade.2C_and_downloading_instructions._Also_set_minimum_php_version.

MatthewVita commented:
Good call!
